src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/unsafe/ForeignUnsafe.java

```diff
@@ -81,4 +83,49 @@ public static long getUnsafeOffset(MemoryAddress address) {
     public static MemorySegment ofNativeUnchecked(MemoryAddress base, long byteSize) {
         return Utils.makeNativeSegmentUnchecked(base, byteSize);
     }
+
+    private static VarHandle arrayHandle(MemoryLayout elemLayout, Class<?> elemCarrier) {
+        return MemoryLayout.ofSequence(1, elemLayout)
+                .varHandle(elemCarrier, MemoryLayout.PathElement.sequenceElement());
+    }
+    private final static VarHandle byteArrHandle = arrayHandle(C_CHAR, byte.class);
+
+    /**
+     * Returns a new native memory segment holding contents of the given Java String
+     * @param str the Java String
+     * @return a new native memory segment
+     */
+    public static MemorySegment makeNativeString(String str) {
+        return makeNativeString(str, str.length() + 1);
+    }
+
+    private static MemorySegment makeNativeString(String str, int length) {
+        MemoryLayout strLayout = MemoryLayout.ofSequence(length, C_CHAR);
+        MemorySegment segment = MemorySegment.allocateNative(strLayout);
+        MemoryAddress addr = segment.baseAddress();
+        for (int i = 0 ; i < str.length() ; i++) {
+            byteArrHandle.set(addr, i, (byte)str.charAt(i));
+        }
+        byteArrHandle.set(addr, (long)str.length(), (byte)0);
+        return segment;
+    }
+
+    /**
+     * Returns a Java String from the contents of the given C '\0' terminated string
+     * @param addr the address of the C string
+     * @return a Java String
+     */
+    public static String toJavaString(MemoryAddress addr) {
+        StringBuilder buf = new StringBuilder();
```

Comment:
There are use cases to take a C string and make it a byte array in nio, and we have to retain the encoding of byte array, some pointed out to me that toJavaString(addr).getBytes() may not be proper if encoding is not properly specified(not my experience though).

Thus I have couple functions like following,
https://github.com/slowhog/panama-foreign/blob/eb79b8c14cfcdeae6a869d6f434929f44ab73dd5/src/java.base/unix/classes/sun/nio/FFIUtils.java#L176

Reply:
Will add Charset accepting variant as well.